### PR TITLE
kubeadm: make the scheduler and KCM connect to the local API endpoint

### PR DESCRIPTION
**Pull request by neolit123**
**What this PR does / why we need it**:

Pinning the kube-controller-manager and kube-scheduler kubeconfig files
to point to the control-plane-endpoint can be problematic during
immutable upgrades if one of these components ends up contacting an N-1
kube-apiserver:
https://kubernetes.io/docs/setup/release/version-skew-policy/#kube-controller-manager-kube-scheduler-and-cloud-controller-manager
For example, the components can send a request for a non-existing API
version.

Instead of using the CPE for these components, use the LocalAPIEndpoint.
This guarantees that the components would talk to the local
kube-apiserver, which should commonly be the same version, unless the user
explicitly patched manifests.

**Which issue(s) this PR fixes**:

xref kubernetes/kubeadm#2271

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm: make the kubeconfig files for the kube-controller-manager and kube-scheduler use the LocalAPIEndpoint instead of the ControlPlaneEndpoint. This makes kubeadm clusters more reseliant to version skew problems during immutable upgrades: https://kubernetes.io/docs/setup/release/version-skew-policy/#kube-controller-manager-kube-scheduler-and-cloud-controller-manager
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go
Comment (neolit123) [inline cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:365]: @fabriziopandini alternatively i can make GetControlPlaneEndpoint() return `controlPlaneEndpoint, localAPIEndpoint, error`

File: cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:370]: What about implementing `kubeadmutil.GetLocalAPIEndpoint` so we are consistent with the same path used above?

File: cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:365]: I'm ok with the solution you are proposing.
Having `GetControlPlaneEndpoint` returning two endpoints seems a little bit weird to me...

**Review by fabriziopandini — COMMENTED** [review]
@neolit123 many thanks for jumping on this issue
Overall LGTM, only one minor nit but not blocking
/cc @ereslibre @rosti

File: cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go
Comment (neolit123) [reply cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:370]: SGTM, i will add this extra utility function.

**Comment by fabriziopandini** [comment]
Test failures seem unrelated to the PR...

File: cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go
Comment (yastij) [inline cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:385]: for control plane nodes, it might make sense for the kubelet to speak to the local API server (e.g. dynamic environments, machines that are using DHCP where the lease expire)

**Review by yastij — COMMENTED** [review]
lgtm

File: cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:385]: From the issue I understood this is going to be tackled in a separated PR, is this right @neolit123

Comment (neolit123) [reply cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:385]: that is true and we spoke about this yesterday.

the problem is that kubeadm join has a kubelet-start phase that blocks until it connects to an api-server at the CPE.
https://github.com/kubernetes/kubernetes/blob/f2e3154a140dc192760a5cf8c01f8044e1aa867b/cmd/kubeadm/app/cmd/phases/join/kubelet.go#L99

https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-join/
the CP manifests + etcd for this node are written after the TLS bootstrap.

this implies rework of the kubeadm join phases and needs some discussion (maybe tomorrow @ office hours).

**Review by ereslibre — APPROVED** [review]
I'm +1 to this -- thank you @neolit123, it's a safe approach reducing the moving parts specially when upgrading.

/lgtm
/hold
^ for other reviews, feel free to unhold.

File: cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go
Comment (yastij) [reply cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:385]: sounds good, I'll join

Comment (neolit123) [reply cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:385]: i will make this PR not close the issue and we can have multiple checkboxes in it to leave time for investigating the kubelet.conf optional change.

Comment (neolit123) [reply cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:385]: done.

Comment (detiber) [reply cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:385]: +1 to both using the local address for the kubelet and also treating it as a separate PR, this will help to greatly simplify the bootstrapping process when using a load balancer, since the load balancer will not need to be pre-created for bootstrapping.

**Comment by timothysc** [comment]
/approve 

@neolit123 What layers of testing do we need to ensure we find these issues earlier?

File: cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go
Comment (neolit123) [reply cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go]: updated.

**Comment by neolit123** [comment]
@timothysc 

> @neolit123 What layers of testing do we need to ensure we find these issues earlier?

node join skew testing:
https://github.com/kubernetes/kubeadm/issues/1661

**Comment by justaugustus** [comment]
Also, this needs re-/lgtm.

**Comment by justaugustus** [comment]
FYI ^^ @kubernetes/ci-signal @hasheddan

**Comment by BenTheElder** [comment]
/retest
https://github.com/kubernetes/test-infra/issues/19080

**Comment by BenTheElder** [comment]
https://github.com/kubernetes/test-infra/issues/19080#issuecomment-685191653
/retest

**Review by rosti — APPROVED** [review]
Thanks @neolit123 !
/lgtm

**Comment by ncdc** [comment]
Are we ok to remove the hold?

**Comment by neolit123** [comment]
yes, i will send a cherry pick later today.

**Comment by neolit123** [comment]
cherry pick is here: https://github.com/kubernetes/kubernetes/pull/94442

**Comment by zhangguanzhang** [comment]
This breaks the rules, the `controlPlaneEndpoint` maybe a domain, if this is a domain, so it will not run ok after your code

**Comment by oldthreefeng** [comment]
when use method `CreateJoinControlPlaneKubeConfigFiles`  with  controlPlaneEndpoint like ` apiserver.cluster.local` to generate config files.  and use  `kubeadm init --config=/root/kubeadm-config.yaml --upload-certs -v 5`
the error occurs like  
```
I0910 15:15:54.436430   52511 kubeconfig.go:84] creating kubeconfig file for controller-manager.conf
currentConfig.Clusters[currentCluster].Server:  https://apiserver.cluster.local:6443 
config.Clusters[expectedCluster].Server:  https://192.168.160.243:6443
a kubeconfig file "/etc/kubernetes/controller-manager.conf" exists already but has got the wrong API Server URL
```

File: cmd/kubeadm/app/util/endpoint.go
Comment (jdef) [inline cmd/kubeadm/app/util/endpoint.go:34]: these docs are now out of date w/ respect to how this method works

Comment (neolit123) [reply cmd/kubeadm/app/util/endpoint.go:34]: this can be fixed in the master branch.

Comment (neolit123) [reply cmd/kubeadm/app/util/endpoint.go:34]: @jdef actually...why do you think the docs of the function are no longer valid?

all comments seem guaranteed to me by the unit tests that we have:
```
// GetControlPlaneEndpoint returns a properly formatted endpoint for the control plane built according following rules:
// - If the controlPlaneEndpoint is defined, use it.
// - if the controlPlaneEndpoint is defined but without a port number, use the controlPlaneEndpoint + localEndpoint.BindPort is used.
// - Otherwise, in case the controlPlaneEndpoint is not defined, use the localEndpoint.AdvertiseAddress + the localEndpoint.BindPort.
```

https://github.com/neolit123/kubernetes/blob/d9441906c4155173ce1a75421d8fcd1d2f79c471/cmd/kubeadm/app/util/endpoint_test.go#L31-L196

**Pull request merged**
